Editing Toolkit / Patterns: Support skipping registering patterns with a requires-align-wide tag

andrewserong opened the pull request:
This PR is designed to skip registering patterns that require full width support, when the site's currently active theme doesn't support full width. It will help avoid the issues described in #47565 and #46215.

#### Changes proposed in this Pull Request

* When registering block patterns via the API, parse the pattern's tags, and for any tags beginning with `requires-`, use the remainder of the tag to look up theme features. If the theme does not support the feature, then skip registering the pattern.
* For example, we should skip registering patterns with the tag `requires-align-wide` on themes that don't support `align-wide`.

#### Screenshots

In a theme that supports `align-wide` we should still see the `images-2` pattern under the Gallery category in the inserter:

The following screenshot is from a test site using the Maywood theme, which has support for `align-wide`:

![image](https://user-images.githubusercontent.com/14988353/99755006-c733eb00-2b3d-11eb-8cfc-208f17b1caa8.png)

In a theme that _does not_ support `align-wide` we should not be able to see the `images-2` pattern under the Gallery category in the inserter.

The following screenshot is from a test site using the Hemingway Rewritten theme, which does _not_ have support for `align-wide`. Note that in the screenshot I'm highlighting the pattern _below_ the one in the above screenshot, showing that it no longer appears:

![image](https://user-images.githubusercontent.com/14988353/99755092-f6e2f300-2b3d-11eb-96c0-369f47adebfa.png)

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Before we can test this, the tag needs to be added to the appropriate patterns. I'll add the tag once we have consensus on the approach. The approach is described in more detail in #47565 — Update: I have added the `requires-align-wide` tag to the main pattern from the original issue (`images-2`).

##### How to test

1. In `apps/editing-toolkit` run `yarn dev --sync` to sync to your sandbox
2. In your sandbox's `0-sandbox.php` file add the following to ensure you're using fresh results from the patterns API that include the appropriate tags:

```
define( 'WP_DISABLE_PATTERN_CACHE', true );
```
3. With your test site sandboxed, and using a theme with `align-wide` support (e.g. any of the suggested themes / Varia child themes such as Maywood), go to edit a post or page, click the inserter and in the Patterns tab, under the Gallery category you should see the pattern with the title "Images" and the text within the pattern will include "5am, the beach". It should look like the top-most screenshot in this PR.

4. Switch your theme to one that does not support `align-wide` (e.g. Hemingway Rewritten) and open up a post. In the inserter, under the Gallery category, you should _not_ be able to see the pattern in question ("Images")

Fixes #47565

andrewserong committed:
Support skipping registering patterns with a requires_align_wide tag

ramonjd commented on apps/editing-toolkit/editing-toolkit-plugin/block-patterns/class-block-patterns-from-api.php:
```diff
@@ -43,6 +50,10 @@ private function __construct() {
+		if ( ! get_theme_support( 'align-wide' ) ) {
+			$this->skip_full_width_patterns = true;
+		}
+
```
I think this is good solution to the problem at hand!

Only that I'm wondering we could be more explicit in the var name, e.g., 

```
$this->supports_align_wide= false !== get_theme_support( 'align-wide' );

```

Not a big deal at all, just semantics and probably subjective at that :) Plus the fact that `can_register_pattern()` tells us all we need to know about "skipping".

And then 
```
	private function can_register_pattern( $pattern ) {
		if ( ! $this->has_align_wide_support && in_array( 'requires_align_wide', array_keys( $pattern['tags'] ), true ) ) {
			return false;
		}

....

```

andrewserong replied:
Nice, thanks for the feedback @ramonjd, I'll update the var name to be more specific, I like `$this->has_align_wide_support` 👍

andrewserong replied:
Abandoning using this variable name, as I've gone with the solution suggested below to infer the feature name from the tag slug.

ramonjd commented on apps/editing-toolkit/editing-toolkit-plugin/block-patterns/class-block-patterns-from-api.php:
```diff
@@ -148,4 +161,18 @@ private function get_block_patterns_locale() {
+	 * @return bool
+	 */
+	private function can_register_pattern( $pattern ) {
+		if ( $this->skip_full_width_patterns && in_array( 'requires_align_wide', array_keys( $pattern['tags'] ), true ) ) {
```
Question: if we're expecting a _specific_ tag anyway, might a future optimization be (assuming they'll be more of such theme exceptions) something like `requires_{ $feature }`, and then

```
	private function can_register_pattern( $pattern ) {
        foreach( $pattern['tags'] as $tag ) {
            $exploded_tag = explode( 'requires_', $tag );
            if ( $exploded_tag[1] ) {
                return false !== get_theme_support( $exploded_tag[1] )
            }
        }
		return true;
	}
```

Very willing to be accused of over-thinking things! :D

andrewserong replied:
@ramonjd I actually quite like that. I'll have a play with it now, because it seems fairly safe and will also future proof it for any similar situations (the issue can be resolved by adding / removing tags, rather than requiring further code changes). Initially I thought it'd be better to cache the `get_theme_support` feature rather than call it on every pattern (since we'll have > 100 patterns), but from looking at how `get_theme_support` [is implemented](https://developer.wordpress.org/reference/functions/get_theme_support/), it's an associative array lookup rather than looping over anything, so seems like it shouldn't be too expensive.

It just means we'd use the tag `requires-align-wide` instead of `requires_align_wide` so that we're hyphenating correctly.

ramonjd replied:
> Initially I thought it'd be better to cache the get_theme_support feature rather than call it on every pattern (since we'll have > 100 patterns),


Ah yeah, good point. That totally escaped me. 

Anyway, it's just an idea seeing as we're expecting a specific string regardless. 

Sorry if I threw an implementation pepper into the soup!

andrewserong replied:
No, no, it's a good idea, thank you! Better to spend a little bit longer getting it right now, so that we can save some time further down the track 😀

andrewserong committed:
Infer the feature to check from the tag slug if it has the 'requires-' prefix

andrewserong commented:
Thanks for the feedback @ramonjd! I've updated this to use the simplified approach you suggested that also supports arbitrary features based on the tag's slug. I went with `preg_split` instead of `explode` so that we can match on the beginning of the slug.

I've also updated the patterns to use a `requires-align-wide` tag for the main problematic pattern, so this PR can now be tested, so it is now ready for a review 😀

ramonjd reviewed (approved):
Nice job! Works according to test instructions

**Maywood**
Pattern is full-width

<img width="1325" alt="Screen Shot 2020-11-23 at 11 03 10 am" src="https://user-images.githubusercontent.com/6458278/99921206-f5ebd480-2d7c-11eb-8f8d-7c4981c9cc2f.png">


**Hemingway Rewritten**
Pattern does not appear in the Patterns Inserter
![Nov-23-2020 11-05-59](https://user-images.githubusercontent.com/6458278/99921210-f8e6c500-2d7c-11eb-8d10-58f2470d235f.gif)
